failure_detector.go

```diff
@@ -0,0 +1,110 @@
+package failure_detector
+
+import (
+	"context"
+	"time"
+
+	batchqueue "github.com/p0lyn0mial/batch-working-queue"
+	ttlstore "github.com/p0lyn0mial/ttl-cache"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// failureDetector is receiving endpoint samples and maintains endpoint status according to logic implemented by a policy evaluator
+// - TODO: internal state is replicated to external storage (atomic.Value) for probing via XYZ() method
+type failureDetector struct {
+	// endpointSampleKeyFn maps collected sample (EndpointSample) for a service to the internal store
+	endpointSampleKeyFn KeyFunc
+
+	//processor retrieves EndpointSamples from the exposed channel and calls out to processBatch() function for processing
+	processor *processor
+
+	// store holds WeightedEndpointStatusStore (samples) per service (namespace/service)
+	store map[string]WeightedEndpointStatusStore
+
+	// createStoreFn a helper function for creating the WeightedEndpointStatusStore store
+	createStoreFn NewStoreFunc
+
+	// policyEvaluatorFn an external policy function for assessing the endpoints
+	policyEvaluatorFn EvaluateFunc
+}
+
+func NewDefaultFailureDetector() *failureDetector {
+	createNewStoreFn := func(ttl time.Duration) WeightedEndpointStatusStore {
+		return newEndpointStore(ttlstore.New(ttl, clock.RealClock{}))
+	}
+	queue := newEndPointSampleBatchQueue(batchqueue.New())
+	return newFailureDetector(EndpointSampleToServiceKeyFunction, SimpleWeightedEndpointStatusEvaluator, createNewStoreFn, queue)
+}
+
+func newFailureDetector(endpointSampleKeyKeyFn KeyFunc, policyEvaluator EvaluateFunc, createStoreFn NewStoreFunc, queue endPointSampleBatchQueue) *failureDetector {
+	fd := &failureDetector{}
+	processor := newProcessor(endpointSampleKeyKeyFn, fd.processBatch, queue)
+	fd.processor = processor
+	fd.store = map[string]WeightedEndpointStatusStore{}
+	fd.endpointSampleKeyFn = endpointSampleKeyKeyFn
+	fd.createStoreFn = createStoreFn
+	fd.policyEvaluatorFn = policyEvaluator
+	return fd
+}
+
+// processBatch starts processing the retrieved EndPointSamples
+// first samples are added to the internal store
+// then it calls out to external policy function for assessing
+// finally it propagates the changes to external read-only store
+func (fd *failureDetector) processBatch(endpointSamples []*EndpointSample) {
+	if len(endpointSamples) == 0 {
+		return
+	}
+	batchKey := fd.endpointSampleKeyFn(endpointSamples[0])
+	endpointsStore := fd.store[batchKey]
+	if endpointsStore == nil {
+		endpointsStore = fd.createStoreFn(60 * time.Second)
```

Comment:
ttl should be pass from outside.

60s means that we forget about broken endpoints within this time, right? Shouldn't we not at least have this at least as long as the times the endpoint controller and node controller need to notice failure?

Reply:
that would be 10ish minutes

Reply:
okay - will factor it out.

failures are just one factor, what about cases in which endpoints actually change because of a new rollout?

Reply:
I would expect some synchronous logic for that.